### PR TITLE
store: prefer /etc/containers/storage.conf

### PR DESCRIPTION
**Pull request by giuseppe**
when running in rootful mode, if it is present, prefer the override path /etc/containers/storage.conf instead of using the  default storage.conf provided by the package under the /usr/share/containers/ directory.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

File: types/utils.go [hunk 1]
```diff
@@ -173,6 +173,9 @@ func DefaultConfigFile(rootless bool) (string, error) {
 		return path, nil
 	}
 	if !rootless {
```
[the remark this answers is not in the document]
Comment (danishprakash) [reply types/utils.go:175]: Again, I'm not entirely sure but looks like the behavior you specified might be handled here:
https://github.com/containers/storage/blob/0af0a9945de2244e55bb548a06d87b4491e79743/types/options.go#L87-L99

[the remark this answers is not in the document]
Comment (giuseppe) [reply types/utils.go:175]: I messed up the commit message.  Yes it applies to root

**Review by rhatdan — APPROVED** [review]
LGTM

**Comment by rhatdan** [comment]
@vrothberg @nalind @mtrmac @flouthoc @umohnani8 @saschagrunert PTAL

**Review by mtrmac — COMMENTED** [review]
Doesn’t this need documentation?

---

Does this actually work? At least Podman callers use `podman/pkg/rootless`, which is a simple EUID check, not handling `_CONTAINERS_ROOTLESS_UID`.  I would expect Podman to set `!rootless` inside a user namespace, then.

OTOH c/storage itself, in `DefaultStoreOptionsAutoDetectUID`, does check for `_CONTAINERS_ROOTLESS_UID`.



Is that intentional? If so, how does this work?

Please _write these things down_. What is the intended semantics of this `rootless` parameter? Then we can discuss whether it is being set correctly by callers; without that we have nothing to really discuss.

**Comment by giuseppe** [comment]
`rootless` is an argument to `DefaultConfigFile` so how it is calculated it is outside the scope of this function.

I agree the code that deals with configuration files is a complete mess and we handle it from so many different places that honestly I don't know how it all is supposed to fit.  It took me a while to understand what is going with the issue I was looking at.

What this PR changes is to honor `/etc/containers/storage.conf` when it exists instead of using `/usr/containers/storage.conf` as the base for the default values, because when we do that, we lose all the settings configured in `/etc/containers/storage.conf`.

**Comment by mtrmac** [comment]
> `rootless` is an argument to `DefaultConfigFile` so how it is calculated it is outside the scope of this function.

I mean, depending on what this parameter means, this could either be a correct PR or a clear bug vs. the stated intent.

I can’t see how propagating the current undefined state is sustainable, and I’m personally not interested. (There are enough other people on Cc:, and I won’t stand in the way of merging if they like it.)

**Comment by rhatdan** [comment]
Lets merge and then work on fixing the documentation.  I think the documentation does state that /etc/containers/storage.conf overrides /usr/share/containers/containers.conf.

**Comment by lukasmrtvy** [comment]
Is possible to backport this for 4.4.0 ? /usr/share/containers/ filesystem path is not writable in FCOS and AFAIK theres no other way, how to override storage.conf for API.

**Comment by rhatdan** [comment]
Yes podman 4.4 will get the latest update containers/storage.

**Pull request merged**